WT-15157 Disagg testing: add leader and follower mode to test/checkpoint

ablekhman opened the pull request:
- add `-d` flag to test/checkpoint for disagg modes: leader, follower, mixed
- define few disagg tests
- tidy-up related CMake scripts, add comments

Siddhartha8899 commented on test/checkpoint/CMakeLists.txt:
```diff
@@ -7,52 +7,143 @@ create_test_executable(test_checkpoint
+
+        # Leader mode
+        list(APPEND disagg_variants
+            "test_checkpoint_leader_${variant_name}|-d leader ${variant_args}")
```
I think including 'disagg' in the test name will make it easier to identify which tests are disaggregated.
how about 
`test_checkpoint_disagg_leader_`?

Siddhartha8899 commented on test/checkpoint/CMakeLists.txt:
```diff
@@ -7,52 +7,143 @@ create_test_executable(test_checkpoint
+            "test_checkpoint_leader_${variant_name}|-d leader ${variant_args}")
+
+        # Follower mode
+        list(APPEND disagg_variants
```
Based on Don's comment, I believe we would delete this?

Siddhartha8899 commented on test/checkpoint/CMakeLists.txt:
```diff
@@ -7,52 +7,143 @@ create_test_executable(test_checkpoint
+# Generates test variants for:
+#   - regular test
+#   - leader test; additional argument: -d leader
+#   - follower test; additional argument -d follower
```
is it better to have disagg test under a different sub directory?
`test/checkpoint`
`test/checkpoint_disagg`

Siddhartha8899 commented on test/utility/test_util.h:
```diff
@@ -134,6 +134,7 @@ typedef struct {
     bool absolute_bucket_dir;  /* Use an absolute bucket path when it is a directory */
     bool compat;               /* Compatibility */
     bool disagg_storage;       /* Uses disaggregated storage */
+    bool disagg_mixed_mode;    /* Switching disaggregated storage mode during the test */
```
should we call it `disagg_switch_mode`?

ablekhman replied:
Good point!

Siddhartha8899 commented on test/checkpoint/test_checkpoint.c:
```diff
@@ -361,6 +385,29 @@ wt_connect(const char *config_open)
       config_open == NULL ? "" : ",", config_open == NULL ? "" : config_open);
 
+    /* Configure disaggregated storage if enabled */
+    if (g.opts.disagg_storage) {
```
We should also ensure precise checkpoint is enabled by appending `precise_checkpoint=true` in the conn_config.

ablekhman replied:
This settings already enabled by the `TESTUTIL_ENV_CONFIG_DISAGG` macro (used within `testutil_disagg_storage_configuration` call).

Siddhartha8899 commented:
As discussed, there are multiple tests related to test_checkpoint scattered across different areas. It would be useful to compile a list of these tests and review which ones are important for IB, which are needed until private preview, and which are incompatible with disagg.

ablekhman committed:
WT-15157 Disagg testing: add leader and follower mode to test/checkpoint

- add `-d` flag to test/checkpoint for disagg modes: leader, follower, mixed
- define few disagg tests
- tidy-up related CMake scripts, add comments

Siddhartha8899 commented on test/evergreen.yml:
```diff
@@ -1700,7 +1700,7 @@ tasks:
       - func: "make check directory"
         vars:
-          ctest_extra_args: -LE "long_running" ${ctest_extra_args}
+          ctest_extra_args: -L "^check$" -LE "long_running" ${ctest_extra_args}
```
@ablekhman Why do we put `-L "^check$"` here?

ablekhman replied:
The format of the label is regular expression. If we simply use `check`, then every label that contains `check` as a substring will match. E.g., `test_checkpoint` will match.

Correct regex for the label makes sure that only tests with actual label `check` will match and run.

Siddhartha8899 commented on test/checkpoint/test_checkpoint.c:
```diff
@@ -113,13 +113,37 @@ main(int argc, char *argv[])
+                g.opts.disagg_switch_mode = false;
+                g.opts.disagg_mode = "leader";
+                g.opts.disagg_page_log = "palm";
+            } else if (strcmp(__wt_optarg, "follower") == 0) {
```
Is there value in running just under `follower` mode?

ablekhman replied:
I just keep it there for completeness. In case someone needs to quickly try the `follower` mode, no additional code changes will be required.

Siddhartha8899 reviewed (approved):
LGTM! mod 2 comments, this looks great @ablekhman thanks!!

Siddhartha8899 commented:
Thanks Alex, this seems all ready to be merged 💯
